lib/io/can.hpp

```diff
@@ -1,12 +1,74 @@
 #pragma once
+#include <cstdint>
+#include <memory>
+#include <optional>
+#include <string>
+#include <unordered_map>
+#include <vector>
 
+#include <core/logger.hpp>
 #include <core/types.hpp>
 
+#ifdef __linux__
+#include <linux/can.h>
+#endif
 namespace hyped::io {
 
-class Can {
+#ifdef __linux__
+using CanFrame = can_frame;
+#else
+// structs and values as defined in <linux/can.h>
+struct CanFrame {
+  std::uint32_t can_id; /* 32 bit CAN_ID + EFF/RTR/ERR flags */
+  std::uint8_t can_dlc; /* frame payload length in byte (0 .. CAN_MAX_DLEN) */
+  std::uint8_t __pad;   /* padding */
+  std::uint8_t __res0;  /* reserved / padding */
+  std::uint8_t __res1;  /* reserved / padding */
+  std::array<std::uint8_t, 8> data;
+};
+struct sockaddr_can {
+  std::uint8_t can_family;
+  std::uint8_t can_ifindex;
+  union {
+    /* transport protocol class address info (e.g. ISOTP) */
+    struct {
+      std::uint32_t rx_id, tx_id;
+    } tp;
+    /* reserved for future CAN protocols address information */
+  } can_addr;
+};
+#define PF_CAN 29
+#define AF_CAN 29
+#define CAN_RAW 1
+#endif
+
+class ICanProcessor {
+ public:
+  virtual void processMessage(const io::CanFrame &frame) = 0;
+};
+
+class ICan {
  public:
-  Can();
+  virtual core::Result initialise(const std::string &can_network_interface)      = 0;
+  virtual core::Result send(const CanFrame &message)                             = 0;
+  virtual std::optional<CanFrame> receive()                                      = 0;
+  virtual core::Result listen()                                                  = 0;
+  virtual void addCanProcessor(const std::uint16_t id, ICanProcessor &processor) = 0;
+};
+
+class Can : public ICan {
```

Comment:
I think this class should be `HardwareCan` and in a separate file `hardware_can.[ch]pp`. This avoid people accidentally including `can.hpp` and using `Can` rather than `ICan`.

Reply:
is that rename can.cpp and split `ICan` and `Can` into seperate hpp files?

Reply:
you should end up with three files

- `can.hpp` containing `ICan`
- `hardware_can.hpp` defining `HardwareCan`
- `hardware_can.cpp` implementing `HardwareCan`